base/broadcast.jl

```diff
@@ -11,7 +11,7 @@ using .Base.Cartesian
 using .Base: Indices, OneTo, tail, to_shape, isoperator, promote_typejoin,
              _msk_end, unsafe_bitgetindex, bitcache_chunks, bitcache_size, dumpbitcache, unalias
 import .Base: copy, copyto!, axes
-export broadcast, broadcast!, BroadcastStyle, broadcast_axes, broadcastable, dotview, @__dot__
+export broadcast, broadcast!, BroadcastStyle, broadcast_axes, broadcastable, dotview, @__dot__, @:
```

Comment:
Maybe export+document a renamed `_lazy` as well? Hiding this functionality behind a macro only does not seem right. But the macro is really nice syntax!

Reply:
Thanks for reviewing the code!

I think it's better to avoid overly increasing the API surface.  I don't see what exporting `Lazy` enables you to do other than what is already possible with `@:`.  Also, implementing `Lazy` can be done in a few lines so I don't think `Base` should export it.  (Note that it still makes sense for `Base` to export `@:`, even though it can be done in a few lines, because it let us use a common syntax across Julia ecosystem.)

Reply:
The main advantage of having a macro is nice looking source. 

The cost of macros in general is increased complexity: It is not immediately clear what the AST is, what the types are, etc. Hence, readers of code need to either trust the docs (that don't tell you what they do on the AST), read the macro code, or macro-expand. In this specific case, the macro does not do anything complicated, and could be written as a `lazy.(...)`. I am surely not the only one who tries to use macros sparingly, and I think lazy is important enough that it should be accessible to people who prefer `view` and dot-syntax or even explicit `broadcast` over `@view`, `@.` and the like.

Reply:
I agree on the point that it is ideal to provide non-macro interface for language features whenever appropriate.  It certainly makes sense for _normal_ functions.  However:

1. Not every feature in Julia has "normal syntax" equivalent.  For example, `@inline`, `@inbounds`, `@simd`, etc. do not have a straight-forward non-macro syntax.

2. You can always run `@macroexpand` or `Meta.@lower` to see what is going on in a given macro (or a specialized syntax).  I think macros in Julia are very transparent.

3. `_lazy` does very unusual thing in the evaluation of the broadcasting expression (because that's its purpose).  There _has_ to be something that signals the readers that the evaluation rule is changed, even if they never used this feature.  A macro is the best syntax for signaling this because that's what macros do.

Reply:
Fair enough.